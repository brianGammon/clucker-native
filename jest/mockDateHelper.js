import moment from 'moment';
import '../src/utils/dateHelper';

jest.mock('../src/utils/dateHelper', () => {
  const module = require.requireActual('../src/utils/dateHelper');
  module.nowAsMoment = () => moment.utc('2018-05-15');
  return module;
});
