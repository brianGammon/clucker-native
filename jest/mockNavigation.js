/* @flow */
type ReturnVals = {
  [methodName: string]: {
    [param: string]: string,
  },
};

type MockNavigation = {
  getParam: (param: string, defaultId: string) => string,
};

const mockNavigation = (returnVals: ReturnVals) => ({
  getParam: jest.fn((param: string, defaultId: string) => {
    if (returnVals.getParam && returnVals.getParam[param]) {
      return returnVals.getParam[param];
    }
    return defaultId;
  }),
}: MockNavigation);

export default mockNavigation;
