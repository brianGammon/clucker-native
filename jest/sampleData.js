const sampleData1 = {
  userSettings: {
    key: 'user1',
    data: {
      currentFlockId: 'flock1',
    },
  },
  flocks: {
    data: {
      flock1: {
        name: 'Test Flock 1',
        ownedBy: 'user1',
      },
      flock2: {
        name: 'Test Flock 2',
        ownedBy: 'user2',
      },
    },
  },
  chickens: {
    data: {
      chicken1: {
        name: 'The First Chicken',
      },
      chicken2: {
        name: 'The Second Chicken',
      },
    },
  },
  eggs: {
    data: {
      egg1: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-01',
        modified: '2018-01-12T18:05:54.806Z',
        notes:
          'This was the record setting egg from May 2017, real egg was 66.1g on this day',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg2: {
        chickenId: 'chicken2',
        chickenName: 'Baby',
        damaged: false,
        date: '2017-09-24',
        modified: '2017-10-12T18:05:54.806Z',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg3: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-02',
        modified: '2018-08-02T18:05:54.806Z',
        notes: '',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 68.1,
      },
    },
  },
};

export default sampleData1;
