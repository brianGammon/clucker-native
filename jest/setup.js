/* eslint-disable import/no-extraneous-dependencies */
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

global.shallow = shallow;
global.mount = mount;

jest.mock('Linking', () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  openURL: jest.fn(),
  canOpenURL: jest.fn(),
  getInitialURL: jest.fn(() => Promise.resolve()),
}));
