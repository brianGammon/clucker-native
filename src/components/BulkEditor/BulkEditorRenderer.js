/* @flow */
import React from 'react';
import {
  Container, Content, View, Form,
} from 'native-base';
import moment from 'moment';
import { Field, FieldGroup, FieldControl } from 'react-reactive-form';
import Header from '../Header';
import DatePicker from '../DatePicker';
import FormInput from '../FormInput';

type Props = {
  form: any,
  onSaveForm: () => void,
  onDateChange: (dateString: string) => void,
};

const BulkEditorRenderer = ({ form, onDateChange, onSaveForm }: Props) => (
  <FieldGroup
    control={form}
    render={({ invalid }) => (
      <Container>
        <Header
          title="Bulk Entry"
          cancelButton
          handleSave={onSaveForm}
          saveDisabled={invalid}
        />
        <Content>
          <View padder>
            <Form>
              <Field
                control={form.get('date')}
                strict={false}
                render={({ value }) => (
                  <DatePicker
                    label="Date"
                    value={value}
                    onDateChange={onDateChange}
                    maximumDate={new Date(moment().valueOf())}
                  />
                )}
              />

              <FieldControl
                name="quantity"
                render={formProps => <FormInput {...formProps} />}
                meta={{
                  label: 'Quantity',
                  keyboardType: 'number-pad',
                }}
              />

              <FieldControl
                name="notes"
                render={formProps => <FormInput {...formProps} />}
                meta={{
                  label: 'Notes',
                  multiline: true,
                  numberOfLines: 5,
                  maxLength: 250,
                }}
              />
            </Form>
          </View>
        </Content>
      </Container>
    )}
  />
);

export default BulkEditorRenderer;
