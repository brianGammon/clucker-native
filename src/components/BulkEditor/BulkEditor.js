/* @flow */
import * as React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { FormBuilder, Validators } from 'react-reactive-form';
import BulkEditorRenderer from './BulkEditorRenderer';
import { nowAsMoment } from '../../utils/dateHelper';
import { type Egg, type Navigation } from '../../types';
import { metaTypes, actionTypes } from '../../redux/constants';
import Loading from '../Loading';
import {
  firebaseUpdateRequested,
  firebaseCreateRequested,
} from '../../redux/actions';
import { dateInRangeValidator } from '../../utils/validators';

type Props = {
  inProgress: boolean,
  error: string,
  eggId: string,
  egg: Egg,
  navigation: Navigation,
  defaultDate: string,
  saveForm: (payload: { eggId?: string, data: Egg }) => void,
  clearError: () => void,
};

type State = {
  formReady: boolean,
};

class BulkEditor extends React.Component<Props, State> {
  form = FormBuilder.group({
    date: ['', [Validators.required, dateInRangeValidator]],
    notes: [''],
    quantity: [
      '',
      [Validators.required, Validators.min(1), Validators.pattern(/^\d+$/)],
    ],
  });

  state = { formReady: false };

  componentDidMount() {
    const { defaultDate, egg } = this.props;
    let defaultState = {
      ...this.form.value,
      date: defaultDate,
    };

    if (egg) {
      const { modified, ...rest } = egg;
      defaultState = { ...defaultState, ...rest };
    }
    this.form.controls.date.setValue(defaultState.date);
    this.form.controls.notes.setValue(defaultState.notes);
    this.form.controls.quantity.setValue(defaultState.quantity);
    this.setState({ formReady: true });
  }

  componentDidUpdate(prevProps) {
    const prevInProgress = prevProps.inProgress;
    const { inProgress, error, navigation } = this.props;
    if (!inProgress && prevInProgress && !error) {
      navigation.goBack();
    }
  }

  componentWillUnmount() {
    const { error, clearError } = this.props;
    if (error) {
      clearError();
    }
  }

  onSaveForm = () => {
    const { egg, eggId, saveForm } = this.props;
    const data = {
      ...egg,
      ...this.form.value,
      chickenId: 'unknown',
      bulkMode: true,
      modified: moment().toISOString(),
    };
    const payload = { eggId, data };
    saveForm(payload);
  };

  onDateChange = (dateString: string) => {
    this.form.controls.date.setValue(dateString);
  };

  render() {
    const { navigation, error } = this.props;
    const { formReady } = this.state;
    if (!formReady) {
      return <Loading />;
    }
    return (
      <BulkEditorRenderer
        navigation={navigation}
        form={this.form}
        onSaveForm={this.onSaveForm}
        error={error}
        onDateChange={this.onDateChange}
      />
    );
  }
}

const mapStateToProps = ({ eggs }, { navigation }) => {
  const eggId = navigation.getParam('eggId', null);
  const defaultDate = navigation.getParam('date', null);
  return {
    eggId,
    egg: eggId ? eggs.data[eggId] : {},
    defaultDate: defaultDate || nowAsMoment().format('YYYY-MM-DD'),
    inProgress: eggs.inProgress,
    error: eggs.error,
  };
};

const mapDispatchToProps = dispatch => ({
  saveForm: (payload) => {
    if (payload.eggId) {
      return dispatch(firebaseUpdateRequested(payload, metaTypes.eggs));
    }
    return dispatch(firebaseCreateRequested(payload, metaTypes.eggs));
  },
  clearError: () => dispatch({ type: actionTypes.CLEAR_ERROR, meta: { type: metaTypes.eggs } }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(BulkEditor);
