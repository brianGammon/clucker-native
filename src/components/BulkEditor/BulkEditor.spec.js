import React from 'react';
import configureStore from 'redux-mock-store';
import BulkEditor from '.';
import mockNavigation from '../../../jest/mockNavigation';
import sampleData from '../../../jest/test-data/full.2017-10-to-2018-05.json';
import '../../../jest/mockDateHelper';

describe('BulkEditor component:', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const store = mockStore(sampleData);

  test('Should render connected BulkEditor with default chicken, egg, and date', () => {
    const paramReturned = {
      getParam: {
        chickenId: 'chicken1',
        date: '2018-02-10',
        eggId: '-L4geyAR0BkMrAjiF7NZ',
      },
    };
    const navigation = mockNavigation(paramReturned);
    const wrapper = shallow(
      <BulkEditor store={store} navigation={navigation} />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('Should render connected BulkEditor with no defaults', () => {
    const paramReturned = {};
    const navigation = mockNavigation(paramReturned);
    const wrapper = shallow(
      <BulkEditor store={store} navigation={navigation} />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
