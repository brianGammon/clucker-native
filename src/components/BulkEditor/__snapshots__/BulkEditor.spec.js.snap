// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkEditor component: Should render connected BulkEditor with default chicken, egg, and date 1`] = `
<BulkEditorRenderer
  error=""
  form={
    FormGroup {
      "_onDisabledChange": Array [],
      "_pendingChange": false,
      "_pendingDirty": false,
      "_pendingTouched": false,
      "asyncValidator": null,
      "controls": Object {
        "date": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "2018-02-06",
          "active": false,
          "asyncValidator": null,
          "errors": null,
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "VALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": [Function],
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": Array [
              [Function],
              [Function],
            ],
          },
          "value": "2018-02-06",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
        "notes": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "Chicken1’s first egg in 3 months!! ",
          "active": false,
          "asyncValidator": null,
          "errors": null,
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "VALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": null,
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": null,
          },
          "value": "Chicken1’s first egg in 3 months!! ",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
        "quantity": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "",
          "active": false,
          "asyncValidator": null,
          "errors": Object {
            "required": true,
          },
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "INVALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": [Function],
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": Array [
              [Function],
              [Function],
              [Function],
            ],
          },
          "value": "",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
      },
      "errors": null,
      "get": [Function],
      "getError": [Function],
      "handleSubmit": [Function],
      "hasError": [Function],
      "meta": Object {},
      "patchValue": [Function],
      "pristine": true,
      "reset": [Function],
      "setValue": [Function],
      "stateChanges": Observable {
        "observers": Array [],
      },
      "status": "INVALID",
      "statusChanges": Observable {
        "observers": Array [],
      },
      "submitted": false,
      "touched": false,
      "validator": null,
      "validatorOrOpts": Object {
        "asyncValidators": null,
        "updateOn": null,
        "validators": null,
      },
      "value": Object {
        "date": "2018-02-06",
        "notes": "Chicken1’s first egg in 3 months!! ",
        "quantity": "",
      },
      "valueChanges": Observable {
        "observers": Array [],
      },
    }
  }
  navigation={
    Object {
      "getParam": [MockFunction] {
        "calls": Array [
          Array [
            "eggId",
            null,
          ],
          Array [
            "date",
            null,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": "-L4geyAR0BkMrAjiF7NZ",
          },
          Object {
            "isThrow": false,
            "value": "2018-02-10",
          },
        ],
      },
    }
  }
  onDateChange={[Function]}
  onSaveForm={[Function]}
/>
`;

exports[`BulkEditor component: Should render connected BulkEditor with no defaults 1`] = `
<BulkEditorRenderer
  error=""
  form={
    FormGroup {
      "_onDisabledChange": Array [],
      "_pendingChange": false,
      "_pendingDirty": false,
      "_pendingTouched": false,
      "asyncValidator": null,
      "controls": Object {
        "date": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "2018-05-15",
          "active": false,
          "asyncValidator": null,
          "errors": null,
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "VALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": [Function],
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": Array [
              [Function],
              [Function],
            ],
          },
          "value": "2018-05-15",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
        "notes": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "",
          "active": false,
          "asyncValidator": null,
          "errors": null,
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "VALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": null,
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": null,
          },
          "value": "",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
        "quantity": FormControl {
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingChange": true,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "_pendingValue": "",
          "active": false,
          "asyncValidator": null,
          "errors": Object {
            "required": true,
          },
          "formState": "",
          "get": [Function],
          "getError": [Function],
          "handler": [Function],
          "hasError": [Function],
          "meta": Object {},
          "onBlur": [Function],
          "onBlurChanges": Observable {
            "observers": Array [],
          },
          "onChange": [Function],
          "onFocus": [Function],
          "onValueChanges": Observable {
            "observers": Array [],
          },
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "INVALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": [Function],
          "validatorsOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": Array [
              [Function],
              [Function],
              [Function],
            ],
          },
          "value": "",
          "valueChanges": Observable {
            "observers": Array [],
          },
        },
      },
      "errors": null,
      "get": [Function],
      "getError": [Function],
      "handleSubmit": [Function],
      "hasError": [Function],
      "meta": Object {},
      "patchValue": [Function],
      "pristine": true,
      "reset": [Function],
      "setValue": [Function],
      "stateChanges": Observable {
        "observers": Array [],
      },
      "status": "INVALID",
      "statusChanges": Observable {
        "observers": Array [],
      },
      "submitted": false,
      "touched": false,
      "validator": null,
      "validatorOrOpts": Object {
        "asyncValidators": null,
        "updateOn": null,
        "validators": null,
      },
      "value": Object {
        "date": "2018-05-15",
        "notes": "",
        "quantity": "",
      },
      "valueChanges": Observable {
        "observers": Array [],
      },
    }
  }
  navigation={
    Object {
      "getParam": [MockFunction] {
        "calls": Array [
          Array [
            "eggId",
            null,
          ],
          Array [
            "date",
            null,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": null,
          },
          Object {
            "isThrow": false,
            "value": null,
          },
        ],
      },
    }
  }
  onDateChange={[Function]}
  onSaveForm={[Function]}
/>
`;
